Makefile

```
@@ -24,112 +18,69 @@ HTTPS_GIT := https://github.com/spacemeshos/api.git
 SSH_GIT := ssh://git@github.com/spacemeshos/api.git
+
 # This controls the version of buf to install and use.
-BUF_VERSION := 1.7.0
+BUF_VERSION := 1.26.1
 
 # This controls the version of protoc to install and use.
-PROTOC_VERSION = 21.5
-
-# Version of go protoc tools
-PROTOC_GEN_GO_VERSION = v1.5.2
-PROTOC_GEN_GRPC_GATEWAY_VERSION = v2.11.2
-
-# The files to run protoc on
-PROTOC_INPUTS := $(shell find ./spacemesh -name *.proto)
-
-# The directory to store go builds
-PROTOC_GO_BUILD_DIR := ./release/go
-
-# Plugins string for go builds (must end in ':')
-PROTOC_GO_PLUGINS := plugins=grpc:
-
-# Options string appended to go build command (optional, obviously)
-PROTOC_GO_OPT := --go_opt=paths=source_relative
-
-# Plugins string for grpc-gateway (must end in ':')
-PROTOC_GATEWAY_PLUGINS := logtostderr=true:
-
-# Service configuration file
-PROTOC_GATEWAY_CONFIG := ./spacemesh/v1/api_config.yaml
-
-# Options string appended to grpc-gateway build command (optional)
-PROTOC_GATEWAY_OPT := --grpc-gateway_opt=paths=source_relative,grpc_api_configuration=$(PROTOC_GATEWAY_CONFIG)
-
-### Everything below this line is meant to be static, i.e. only adjust the above variables. ###
+PROTOC_VERSION = 24.3
+
+# Everything below this line is meant to be static, i.e. only adjust the above variables. ###
+
+ifeq ($(OS),Windows_NT)
+	UNAME_OS := windows
+	ifeq ($(PROCESSOR_ARCHITECTURE),AMD64)
+		UNAME_ARCH := x86_64
+	endif
+	ifeq ($(PROCESSOR_ARCHITECTURE),ARM64)
+		UNAME_ARCH := aarch64
+	endif
+	PROTOC_BUILD := win64
+
+	BIN_DIR := $(abspath .)/bin
+	export PATH := $(BIN_DIR);$(PATH)
+	TMP_PROTOC := $(TEMP)/protoc-$(RANDOM)
+else
+	UNAME_OS := $(shell uname -s)
+	UNAME_ARCH := $(shell uname -m)
+	PROTOC_BUILD := $(shell echo ${UNAME_OS}-${UNAME_ARCH} | tr '[:upper:]' '[:lower:]' | sed 's/darwin/osx/' | sed 's/arm64/aarch_64/' | sed 's/aarch64/aarch_64/')
+
+ 	BIN_DIR := $(abspath .)/bin
+ 	export PATH := $(BIN_DIR):$(PATH)
+ 	TMP_PROTOC := $(shell mktemp -d)
+endif
 
-UNAME_OS := $(shell uname -s)
-UNAME_ARCH := $(shell uname -m)
+# `go install` will put binaries in $(GOBIN), avoiding
+# messing up with global environment.
+export GOBIN := $(BIN_DIR)
 
-UNAME_OS_PROTOC := $(shell uname -s | tr '[:upper:]' '[:lower:]' | sed 's/darwin/osx/')
-# Buf will be cached to ~/.cache/buf-example.
-CACHE_BASE := $(HOME)/.cache/$(PROJECT)
-# This allows switching between i.e a Docker container and your local setup without overwriting.
-CACHE := $(CACHE_BASE)/$(UNAME_OS)/$(UNAME_ARCH)
-# The location where buf will be installed.
-CACHE_BIN := $(CACHE)/bin
-# Marker files are put into this directory to denote the current version of binaries that are installed.
-CACHE_VERSIONS := $(CACHE)/versions
-# We do some temporary work here
-CACHE_TMP := $(CACHE_BASE)/tmp
+install-buf:
+	@mkdir -p $(BIN_DIR)
+	curl -sSL "https://github.com/bufbuild/buf/releases/download/v$(BUF_VERSION)/buf-$(UNAME_OS)-$(UNAME_ARCH)" -o $(BIN_DIR)/buf
+	@chmod +x $(BIN_DIR)/buf
+.PHONY: install-buf
```

The existing approach with "cache" has a nice property that it automatically detects if a new version needs to be installed. That's lost now :(

Although it could be ~easily restored. How about:
```suggestion
BUF := $(BIN_DIR)/buf_$(BUF_VERSION)
$(BUF):
	@mkdir -p $(BIN_DIR)
	curl -sSL "https://github.com/bufbuild/buf/releases/download/v$(BUF_VERSION)/buf-$(UNAME_OS)-$(UNAME_ARCH)" -o $@
	@chmod +x $@
```

And then use it like that:

```makefile
build: $(BUF)
	$(BUF) generate
```

Now, buf will be automatically updated (or downloaded if doesn't exist) on `make build` . No need for `make install` at all.

:bulb: It could also remove existing bufs before downloading the new version:
```makefile
	rm $(BIN_DIR)/buf_*
```

---

I wouldn't overcomplicate it. We can do a similar approach as we do with our own dependencies: have the zip file containing buf as a `make` target that downloads and extracts the executable. If the zip file is already there it will be a no-op and otherwise the executable will be updated, wdyt?

---

I did something similar: `buf` is now installed with version appended to `bin/buf_*` and the file is a target of the `Makefile`. Updating the `buf` version will download the new version and delete the previous.

Additionally `protoc` isn't installed any more, it's not actually needed (only its plugins).